packages/babel-plugin-jest-hoist/src/index.ts

```diff
@@ -120,41 +120,45 @@ FUNCTIONS.mock = args => {
     moduleFactory.traverse(IDVisitor, {ids});
     for (const id of ids) {
       const {name} = id.node;
-      let found = false;
       let scope = id.scope;
 
       while (scope !== parentScope) {
         if (scope.bindings[name]) {
-          found = true;
-          break;
+          return true;
         }
 
         scope = scope.parent;
       }
 
-      if (!found) {
-        const isAllowedIdentifier =
-          (scope.hasGlobal(name) && ALLOWED_IDENTIFIERS.has(name)) ||
-          /^mock/i.test(name) ||
-          // Allow istanbul's coverage variable to pass.
-          /^(?:__)?cov/.test(name);
-
-        if (!isAllowedIdentifier) {
-          throw id.buildCodeFrameError(
-            'The module factory of `jest.mock()` is not allowed to ' +
-              'reference any out-of-scope variables.\n' +
-              'Invalid variable access: ' +
-              name +
-              '\n' +
-              'Allowed objects: ' +
-              Array.from(ALLOWED_IDENTIFIERS).join(', ') +
-              '.\n' +
-              'Note: This is a precaution to guard against uninitialized mock ' +
-              'variables. If it is ensured that the mock is required lazily, ' +
-              'variable names prefixed with `mock` (case insensitive) are permitted.\n',
-            ReferenceError,
-          );
-        }
+      const binding = scope.bindings[name];
+
+      // @ts-expect-error `init` does not exist
+      if (binding?.constant && scope.isPure(binding.path.node.init, true)) {
+        // how to hoist???
```

Comment:
@nicolo-ribaudo @jeysal could any of you tell me how to lift this up higher? 😀 This "passes" the test (i.e. it doesn't throw) by doing the early return, but the statement is left in place so it might be `undefined`

Reply:
I was going to suggest using "the `z-index` rule": add a `9` to until it works.

However, it seems that `jest.mock` calls have a priority of `Infinity`:
https://github.com/facebook/jest/blob/1818d84b2cf20ecb8db2215b3ad6b5e397ab2673/packages/babel-plugin-jest-hoist/src/index.js#L166

Can you try replacing `Infinity` with `999` and use `9999` to hoist this one?

Reply:
Or maybe just using `Infinity` in both places work because `var _jsxFileName` was already before the mock call in the original code.

Reply:
That's old code, after #9806 we don't use `_blockHoist` anymore. But I can try to add it in and see what happens 🙂

Reply:
Yep, we got rid of that ugly semi-public API and now do relatively clean traversal and I'd prefer to keep it that way 😅

Reply:
adding `binding.path.node._blockHoist = 999999;` makes no difference to the code that is output

Reply:
FWIW, `binding.path.parent._blockHoist = Infinity;` moved the statement, but only above the `import` - not above the injected, hoisted `jest.mock` statement.

I'm currently leaning towards somehow replacing `var _jsxFileName = 'whatever'` with 

```js
function _jsxFileName() {
  return 'whatever';
}
```

and then replacing usage. That's way beyond what I can do with AST manipulation tho, so hoping one of you are able to find the time 😀 

If not that, then just replacing all the usage of the variable with its value and get rid of the binding entirely. `binding.referenced` is `false` though, so I'm not sure how to figure out where it's referenced and replace?